### PR TITLE
feat(internal/fs/node): Restore ownership by name

### LINKED ISSUE #3572: Also support restoring ownership by name
Output of `restic version`
--------------------------

restic 0.12.0 compiled with go1.16.6 on linux/amd64

What should restic do differently? Which functionality do you think we should add?
----------------------------------------------------------------------------------

It would be great, if `restic restore` would support an option for restoring ownership information from the symbolic user/group names instead of just restoring the uid/gid.

Also, it would be extra great, if restic would warn, by default, if the uid/gids of restored files map to different user/group names on the target system than on the backed up system.

What are you trying to do? What problem would this solve?
---------------------------------------------------------

I was doing a disaster recovery of a system, i.e. I installed it from scratch, installed and configured some services via a playbook and restored some data file with restic. It turns out that many system users such as 'nginx' have dynamically assigned uid/gids on modern Linux distributions. Thus, they likely change if the system is re-deployed from scratch.

Since restic stores/restored raw uids/gids, I had to chown the restored files which was quite tedious.

See also https://forum.restic.net/t/restore-with-numeric-uid-or-with-username/3811/2 for a similar description of this use case.

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------

Restic recently helped me backing up some old Intel Core 1 host with just 2 GB RAM with reasonable speed which was nice.

### PR DESCRIPTION
**Pull request by provokateurin**
What does this PR change? What problem does it solve?
-----------------------------------------------------

When doing a restore the UID and GID for the chown are now looked up on the current system, so that the ownership will actually match the expected user/group names.

I am not sure if this needs extra tests, since the existing ones all pass which at least validates that it didn't break existing behavior.
I am also not sure if this should be behind a CLI flag and if this needs additional documentation.
Please let me know if any of those things are needed, I'm happy to put more work into this feature.

This is my first PR to restic, but I've been a very happy user so far and this issues has been biting me quite a bit. I went so far to even create a script that would be put into every snapshot, in order to correctly restore the ownership, but it would be great to have a solution directly in restic.

Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------

Fixes https://github.com/restic/restic/issues/3572

Checklist
---------

- [ ] I have added tests for all code changes.
- [ ] I have added documentation for relevant changes (in the manual).
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [x] I'm done! This pull request is ready for review.

### CODE REVIEW

**Comment by zumm** [comment]
I think it should be optinal. Setting it as default is breaking change.

Also any chance we could get it for Windows?

**Comment by provokateurin** [comment]
While I can see why you say it is a breaking change, I don't think it is one since the original behavior should stay on most systems where the uid and gid mappings are the same. It just fixes the behavior for systems where this isn't the case.

I don't know about Windows, but it didn't have any chown-like thing before on a restore. Also I just don't know much about Windows in general, do they even have the concept of file ownership with uid and gid?

**Comment by zumm** [comment]
There is concept of Security Descriptors on Windows. Support was introduced in #4509 (and this is the example why features should be optional even if they seems like good default).

**Comment by provokateurin** [comment]
@zumm I added a CLI flag to optionally enable the new behavior.

**Comment by MichaelEischer** [comment]
> There is concept of Security Descriptors on Windows. Support was introduced in #4509 (and this is the example why features should be optional even if they seems like good default).

Windows will have to be handled separately. The security descriptors rely on unique IDs which can't be easily translated between systems.

File: cmd/restic/cmd_restore.go [hunk 1]
```diff
@@ -86,6 +87,7 @@ func (opts *RestoreOptions) AddFlags(f *pflag.FlagSet) {
 	f.BoolVar(&opts.Verify, "verify", false, "verify restored files content")
 	f.Var(&opts.Overwrite, "overwrite", "overwrite behavior, one of (always|if-changed|if-newer|never)")
 	f.BoolVar(&opts.Delete, "delete", false, "delete files from target directory if they do not exist in snapshot. Use '--dry-run -vv' to check what would be deleted")
+	f.BoolVar(&opts.OwnershipByName, "ownership-by-name", false, "restore file ownership by uid and gid")
```
Comment (MichaelEischer) [inline cmd/restic/cmd_restore.go]: That option doesn't work on Windows for now and thus shouldn't be offered there.

File: changelog/unreleased/issue-3572 [hunk 2]
```diff
@@ -0,0 +1,7 @@
+Enhancement: Restore ownership by name
+
+Restic now optionally restores file ownership on UNIX systems by the names of the user and group.
+This allows restoring snapshots on a system where the UID/GID are not the same as they were on the system where the snapshot was created.
+Previously, on a system where the UID/username mapping or the GID/group name mapping did not match the expected values, restoring a snapshot would result in incorrect ownership of the files.
+
+https://github.com/restic/restic/issues/3572
```
Comment (MichaelEischer) [inline changelog/unreleased/issue-3572:8]: Include a link to this PR.

File: changelog/unreleased/issue-3572 [hunk 3]
```diff
@@ -0,0 +1,7 @@
+Enhancement: Restore ownership by name
+
+Restic now optionally restores file ownership on UNIX systems by the names of the user and group.
+This allows restoring snapshots on a system where the UID/GID are not the same as they were on the system where the snapshot was created.
+Previously, on a system where the UID/username mapping or the GID/group name mapping did not match the expected values, restoring a snapshot would result in incorrect ownership of the files.
```
Comment (MichaelEischer) [inline changelog/unreleased/issue-3572]: The "previous behavior" part should come first in the description.

File: internal/fs/node_unix.go [hunk 4]
```diff
@@ -9,8 +9,17 @@ import (
 	"github.com/restic/restic/internal/restic"
 )
 
-func lchown(name string, uid, gid int) error {
-	return os.Lchown(name, uid, gid)
+func lchown(name string, node *restic.Node, lookupByName bool) error {
```
Comment (MichaelEischer) [inline internal/fs/node_unix.go]: This definitely needs a test.

File: changelog/unreleased/issue-3572 [hunk 5]
```diff
@@ -0,0 +1,7 @@
+Enhancement: Restore ownership by name
```
Comment (MichaelEischer) [inline changelog/unreleased/issue-3572]:
```suggestion
Enhancement: Support restoring ownership by name
```

File: changelog/unreleased/issue-3572 [hunk 6]
```diff
@@ -0,0 +1,7 @@
...
+
+Restic now optionally restores file ownership on UNIX systems by the names of the user and group.
```
Comment (MichaelEischer) [inline changelog/unreleased/issue-3572]: Please mention the necessary parameter and that it is for the `restore` command.

**Comment by provokateurin** [comment]
Rebased to resolve conflicts

**Review by MichaelEischer — CHANGES_REQUESTED** [review]
While taking a closer look, I noticed a rather ugly problem that cannot be solved by this PR. Restic stores POSIX ACLs in a backup, but only in their numeric on-disk format. Thus, we have no way of restoring those based on their name.

However, fixing that is out of scope for this PR. Instead let's document the limitation somewhere (ideally in the CLI parameter). Besides that the PR only needs tests for the new codepaths.

File: internal/fs/node.go [hunk 7]
```diff
@@ -130,6 +130,37 @@ func lookupUsername(uid uint32) string {
 	return username
 }
 
+var (
+	userNameLookupCache      = make(map[string]uint32)
+	userNameLookupCacheMutex = sync.RWMutex{}
+)
+
+// Cached uid lookup by user name. Returns 0 when no id can be found.
+func lookupUid(userName string) uint32 {
```
Comment (MichaelEischer) [inline internal/fs/node.go:141]: This will require a `//nolint:revive // captialization is correct as is` comment

File: changelog/unreleased/issue-3572 (hunk 3 above)
Comment (rawtaz) [reply changelog/unreleased/issue-3572]: See https://github.com/restic/restic/blob/master/changelog/TEMPLATE for a guiding example on how to structure the changelog file :)

**Comment by provokateurin** [comment]
> Instead let's document the limitation somewhere (ideally in the CLI parameter)

What exactly would you write there? I feel like it does not really belong into the usage description of the new CLI flag, but maybe for the restore command itself.

File: internal/fs/node_unix.go (hunk 4 above)
Comment (provokateurin) [reply internal/fs/node_unix.go]: I tried that, but unfortunately chown needs root permission, so it would add the requirement to run the tests with some kind of root access which is a really bad idea.
What do you think is the right way to test this?

Comment (MichaelEischer) [reply internal/fs/node_unix.go]: We definitely don't want to require running the tests as root. But a basic sanity check should be possible. A user is is allowed to chown files owned by them to their own uid. This is effectively a no-op, but should be enough to sanity check the overall username / uid translation steps.

Other than that the tests in `internal/fs/node_test.go` should also be run with `lookupByName=true` at least once. There will be quite a few corner cases which won't be covered by those tests, but a basic smoke test is better than nothing at all.

**Comment by MichaelEischer** [comment]
> What exactly would you write there? I feel like it does not really belong into the usage description of the new CLI flag, but maybe for the restore command itself.

The CLI flag could get a `(except POSIX ACLs)` suffix. Or add a somewhat longer description to the restore command itself. Or probably better: add both.

File: cmd/restic/cmd_restore.go [hunk 8]
```diff
@@ -84,6 +86,9 @@ func (opts *RestoreOptions) AddFlags(f *pflag.FlagSet) {
 	f.BoolVar(&opts.Verify, "verify", false, "verify restored files content")
 	f.Var(&opts.Overwrite, "overwrite", "overwrite behavior, one of (always|if-changed|if-newer|never)")
 	f.BoolVar(&opts.Delete, "delete", false, "delete files from target directory if they do not exist in snapshot. Use '--dry-run -vv' to check what would be deleted")
+	if runtime.GOOS != "windows" {
+		f.BoolVar(&opts.OwnershipByName, "ownership-by-name", false, "restore file ownership by uid and gid")
```
Comment (MichaelEischer) [inline cmd/restic/cmd_restore.go]:
```suggestion
		f.BoolVar(&opts.OwnershipByName, "ownership-by-name", false, "restore file ownership by username and groupname")
```

**Comment by provokateurin** [comment]
@MichaelEischer anything missing from my side or something I can do to get it merged? No pressure though, I see you are doing a lot of work for restic already! :heart:

**Comment by provokateurin** [comment]
Sorry to ping again, @MichaelEischer any chance you could have another look? I resolved all review comments.

File: changelog/unreleased/issue-3572 [hunk 9]
```diff
@@ -0,0 +1,8 @@
+Enhancement: Support restoring ownership by name
+
+Restic restore used to restore file ownership on UNIX systems by UID and GID.
+It now allows restoring the file ownership by user name and group name with `--ownership-by-name`.
+This allows restoring snapshots on a system where the UID/GID are not the same as they were on the system where the snapshot was created.
```
Comment (MichaelEischer) [inline changelog/unreleased/issue-3572:5]: Should also mention that this does not work for POSIX ACLs so far.

File: changelog/unreleased/issue-3572 [hunk 10]
```diff
@@ -0,0 +1,8 @@
+Enhancement: Support restoring ownership by name
```
Comment (MichaelEischer) [inline changelog/unreleased/issue-3572]:
```suggestion
Enhancement: Support restoring ownership by name on UNIX systems
```

**Review by MichaelEischer — CHANGES_REQUESTED** [review]
Sorry for the long delay. I only have two small nits regarding the changelog left, then this is ready for merging.

**Comment by provokateurin** [comment]
Done! Thanks for reviewing!

**Review by MichaelEischer — APPROVED** [review]
LGTM.

For handling POSIX ACLs correctly, I've created a minimal issue: https://github.com/restic/restic/issues/5598
